internal/rm/rm.go

```diff
@@ -110,6 +112,35 @@ func (r *resourceManager) Devices() Devices {
 	return r.devices
 }
 
+var errInvalidRequest = errors.New("invalid request")
+
+// ValidateRequests checks the requested IDs against the resource manager configuration.
+// This asserts that all requested IDs are known to the resource manager and that the request is
+// valid for a specified sharing configuration.
+// specified sharing configuration is respected.
+func (r *resourceManager) ValidateRequest(ids AnnotatedIDs) error {
+	for _, id := range ids {
+		if !r.devices.Contains(id) {
+			return fmt.Errorf("%w: unknown device: %s", errInvalidRequest, id)
+		}
+	}
+	// If the devices being allocated are replicas, then (conditionally)
+	// error out if more than one resource is being allocated.
+	includesReplicas := ids.AnyHasAnnotations()
+	numRequestedDevices := len(ids)
+	switch r.config.Sharing.SharingStrategy() {
+	case spec.SharingStrategyTimeSlicing:
+		if includesReplicas && numRequestedDevices > 1 && r.config.Sharing.ReplicatedResources().FailRequestsGreaterThanOne {
+			return fmt.Errorf("%w: maximum request size for shared resources is 1; found %d", errInvalidRequest, numRequestedDevices)
+		}
+	case spec.SharingStrategyMPS:
+		if includesReplicas && numRequestedDevices > 1 {
+			return fmt.Errorf("%w: maximum request size for shared resources is 1; found %d", errInvalidRequest, numRequestedDevices)
+		}
+	}
```

Comment:
Do we actually need this switch anymore since you now do this early on?:
```
	if config.Sharing.MPS != nil {
		config.Sharing.MPS.FailRequestsGreaterThanOne = true
	}
```

Reply:
It's not strictly necessary, but this switch makes it simpler to add extensions where we allow more flexibility for MPS. It also makes it clear that that the behaviour for MPS *may* be different to timeslicing.

It also ensures that modifications to the config (after we set the value as you point out above) doesn't change the behaviour.

I'm willing to change it if we think that we'd rather not have this now.

Reply:
OK; since they are identical for now, can we at least change it to:
```
switch r.config.Sharing.SharingStrategy() {
	case spec.SharingStrategyTimeSlicing:
		if includesReplicas && numRequestedDevices > 1 && r.config.Sharing.ReplicatedResources().FailRequestsGreaterThanOne {
			return fmt.Errorf("%w: maximum request size for shared resources is 1; found %d", errInvalidRequest, numRequestedDevices)
	case spec.SharingStrategyMPS:
		if includesReplicas && numRequestedDevices > 1 && r.config.Sharing.ReplicatedResources().FailRequestsGreaterThanOne {
			return fmt.Errorf("%w: maximum request size for shared resources is 1; found %d", errInvalidRequest, numRequestedDevices)
}
```
or
```
switch r.config.Sharing.SharingStrategy() {
	case spec.SharingStrategyTimeSlicing:
	case spec.SharingStrategyMPS:
		if includesReplicas && numRequestedDevices > 1 && r.config.Sharing.ReplicatedResources().FailRequestsGreaterThanOne {
			return fmt.Errorf("%w: maximum request size for shared resources is 1; found %d", errInvalidRequest, numRequestedDevices)
}
```
Meaning, keep the switch, but keep the logic itself the same.

Reply:
I'm willing to make the change, but would like to make an additional case against it. I thought we agreed that the `failRequestsGreaterThanOne` and the semantics around it is a workaround for what should have been the correct default behaviour for timeSlicing. I would rather *not* carry this forward into the MPS implementation. Ideally I would have introduced a different type for the MPS `ReplicatedResources`, but deemed that reusing the existing defintion would be simpler in terms of the affect on other labelling code.

Note that we are setting the value to `true` for cosmetic reason -- so that the user doesn't see `failRequestsGreaterThanOne: false` when the config is printed to the logs -- and should not affect the behaviour. Once we extend the MPS API to be richer (we talked about supporting something like mixed mig profiles) we would replace the data structure and no longer need to deal with this flag directly.

Reply:
Honestly, I don't feel too strongly either way. Maybe having a comment to the effect of what you wrote above would help. Otherwise, it appears to the casual reader of the code as if we made a mistake, and overlooked the fact that we *should* be doing the exact same operation in both places.